### ISSUE TITLE
Bad / misleading error message with auto deref and mutable borrows of multiple fields

### ISSUE THREAD

**Issue by TimNN** [rust-lang/rust#32930]
Consider ([playpen](https://play.rust-lang.org/?gist=c948d8a0c4e8e0bc697231d743c4f4bd&version=stable&backtrace=0)):

``` rust
struct Foo {
    a: u32,
    b: u32,
}

struct Container {
    foo: Foo,
}

impl ::std::ops::Deref for Container {
    type Target = Foo;

    fn deref(&self) -> &Foo { &self.foo }
}

impl ::std::ops::DerefMut for Container {
    fn deref_mut(&mut self) -> &mut Foo { &mut self.foo }
}

fn main() {
    let mut c = Container { foo: Foo { a: 0, b: 0 } };

    {
        // let mut c = &mut *c; // uncomment to make it work
        let _a = &mut c.a;
        let _b = &mut c.b; // ERROR!
    }

    let mut d = Box::new(Foo { a: 0, b: 0 });

    {
        // let mut d = &mut *d; // uncomment to make it work
        let _a = &mut d.a;
        let _b = &mut d.b; // ERROR!
    }
}
```

Intuitively this should work since rust allows mutable borrows of disjoint fields however it does not because auto deref is involved.

The error message here is particularly bad / misleading for the `Box` case, in my opinion:

``` plain
<anon>:34:23: 34:26 error: cannot borrow `d` (here through borrowing `d.b`) as mutable more than once at a time [E0499]
<anon>:34         let _b = &mut d.b;
                                ^~~
<anon>:34:23: 34:26 help: see the detailed explanation for E0499
<anon>:33:23: 33:26 note: previous borrow of `d` occurs here (through borrowing `d.a`); the mutable borrow prevents subsequent moves, borrows, or modification of `d` until the borrow ends
<anon>:33         let _a = &mut d.a;
                                ^~~
<anon>:35:6: 35:6 note: previous borrow ends here
<anon>:31     {
<anon>:32         // let mut d = &mut *d; // uncomment to make it work
<anon>:33         let _a = &mut d.a;
<anon>:34         let _b = &mut d.b;
<anon>:35     }
```

First found on [reddit](https://www.reddit.com/r/rust/comments/4eknip/cannot_implement_basic_data_structures_in_rust/).

**Comment by TimNN** [comment]
Related to #30104

**Comment by Pauan** [comment]
I just ran into this with `RefCell` + `borrow_mut` (which in my opinion is a common situation). Thanks a lot for the `&mut *foo` workaround!

**Comment by shepmaster** [comment]
The original example **compiles successfully** with NLL 🎉 .

**Comment by pnkfelix** [comment]
NLL (migrate mode) is enabled in all editions as of PR #59114. Verified that test case compiled in Nightly 2015 edition; closing as fixed.
